pkg/api/batch/warnings.go

```diff
@@ -0,0 +1,38 @@
+package batch
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/kubernetes/pkg/apis/batch"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+var TTLWarningMsg = "This job doesn't have TTLAfterFinished field and not managed by a higher level controller. " +
+	"Setting TTLAfterFinished on job.spec helps in removal of dependant pods"
```

Comment:
The threshold for returning warnings to the user has been to alert them to flaws in their requests or use of deprecated things. It is not clear to me that someone creating a job without a `ttlSecondsAfterFinished` set is doing something that deserves a warning. If they are intentionally doing that, and are managing job cleanup themselves separately, this will be very annoying and not really actionable.

While we think about whether a warning should be issued at all, a few comments about the specific text here:
1. Use API field names, not go struct field names (and lead with the field in question, `spec.ttlSecondsAfterFinished is unset; ...`)
2. I think the primary reason to set this is not to help with pod cleanup, but to help with job cleanup (which has a side effect of cleaning up the pods)
3. This is pretty verbose, something shorter would be better

Reply:
>  It is not clear to me that someone creating a job without a ttlSecondsAfterFinished set is doing something that deserves a warning. If they are intentionally doing that, and are managing job cleanup themselves separately, this will be very annoying and not really actionable.


I agree with your assessment. However, don't you think they are cleaning up lingering pods because of the [issue](https://github.com/kubernetes/kubernetes/pull/103449/files#diff-c6dd6e4d6de95f2b10588edc34d31d2500af39235fc799cfe3c85301c6a4c289L65) and if given a chance they'd prefer automatic clean up? I know that this warning will be useless in those scenarios. Can't we warn in potential failure scenarios? Do we always warn if it is actionable?

Reply:
> However, don't you think they are cleaning up lingering pods because of the [issue](https://github.com/kubernetes/kubernetes/pull/103449/files#diff-c6dd6e4d6de95f2b10588edc34d31d2500af39235fc799cfe3c85301c6a4c289L65) and if given a chance they'd prefer automatic clean up?

That's unknowable... if they have a post-completion process that requires the job to exist and then cleans up the job, not having a ttl is reasonable. Currently the guidance on warnings is that they only be emitted when we are sure the usage is incorrect / deprecated.

Reply:
So, how to inform  the user of a scenario where there can be potential implication of the choices they make?

Reply:
that's what the warning on job *deletion* was for

Reply:
But that's way down in the stack where we are sure that this deletion causes lingering pods and it is irreversible at that point. Can't we warn early during the creation itself that there is a potential problem and you can update it to make sure that you won't face any problems in future?

Reply:
I think warning about potential problems that may not apply will lead to people learning to ignore warnings, so I'm not in favor of that.

Reply:
Ok. I see your point. I am fine with not having the warning which is not always actionable. What other mechanisms or tools do we have to ensure that these type of `potential problems` are properly surfaced to the end-user  so that they can make informed decisions, apart from documenting the behavior?

Reply:
documentation and metrics seem like the main possibilities

Reply:
I spoke with Ravi about it on slack, and I agree with Jordan, that the best way forward with this is to start with a documentation updates recommending setting `ttlSecondsAfterFinished`. I share the same concerns like Jordan about too many warning leading to people ignoring those.

Reply:
Fair enough, I am closing the PR now. Will open a docs PR. @soltysh and I also discussed about metrics and potential alerts and I'll go through existing metrics if they're already available to be used.